### ISSUE TITLE
Bikeshed: render-subtree tokens.

### ISSUE THREAD

**Issue by vmpstr** [WICG/display-locking#110]
Hi,

We currently have `render-subtree: invisible skip-viewport-activation skip-activation` Historically, these token names made sense!

Now? I'm not sure.

`render-subtree: invisible` manages locking/unlocking automatically and never sends any signals. Should it be... `render-subtree: managed`? `render-subtree: auto`? auto is bad since it's a common 'default'. What should that be named?

`render-subtree: invisible skip-activation` doesn't manage anything. Doesn't allow searchability with in, it's essentially a state-preserving display: none (for the subtree). What should we call this? I propose `render-subtree: invisible`, because... well that's what it is: invisible.

`render-subtree: invisible skip-viewport-activation` is kind of like invisible but allows searchability within which is good if we want script to uninvisible this. Collapsed but searchable sections is a use case. I propose `render-subtree: invisible searchable`. I don't like this last one, but I don't know what else we would call it.

It's conceivable that we might not need tokens if we have a single word for the last state, and just have an enum. However, with hold-loads and hold-upgrades and things like that, we might still need tokens. Or, maybe hold-loads and hold-upgrades shouldn't fall into a bucket of render-subtree and be something else entirely? Not sure.

Can we please come up with a set of good names for the states that we do have, with a possibility of extending this for loads/upgrades in the future?

/cc @cbiesinger @chrishtr @fergald @flackr @rakina

**Comment by fergald** [comment]
So this is kind of about whether we want to provide nuts and bolts that can
be used in ways that might surprise us (or might surprise the developer) vs
higher-level concepts. If we go with the higher level concepts, I would
suggest something like

- `auto`: please optimize this for me, this content is in the document,
make it appear when it should appear
- `managed`; I'm clever, I'm managing these, I will control their
visibility but this content is in the document so *do* send me events for
find in page, tab navigation, select, maybe even scrolling etc and maybe
even unhide content if my event handler screws up (the way we talked about
for SID).
- `hidden` / `preparing`: this content is not in the document, I'm getting
it ready to swap it in

I'm tempted to say that `auto` and `managed` are modifiers and the element
will flip between `auto` and `auto hidden` depending on whether the browser
thinks it should be hidden or not and between `managed` and `managed
hidden` under the control of JS. And maybe even say that trying to set
`hidden` directly will just not work with `auto`. This also makes me think
that they 2 separate attributes, `renderSubtreeMode="auto"|"managed"|""`,
`renderSubtreeVisible=true/false` and `renderSubtreeVisible` becomes
effectively read-only when `renderSubtreeMode="auto"`

I wonder does that break anything from the CSS POV?

**Comment by tabatkins** [comment]
I think I'm in agreement with Vlad that `auto` doesn't communicate the right thing here; we usually reserve that keyword for initial values that have to mean "do some smart/complex thing by default for me, since I'm not giving any more specific instruction".

Stream of consciousness incoming~

-----

So just to lay them out clearer for myself, we have four states currently:

1. normal rendering
2. completely hidden; "state-preserving display:none" as Vlad puts it
3. hidden but the UA can interact with it and will inform the author when it should be unlocked
4. hidden but the UA can interact and auto-unlocks it when necessary

I think we can start with a pair of "normal" (for 1) and "hidden" (for 2-4), since that's the major distinction to draw here. (Or "invisible", but 'visibility' already uses 'visible'/'hidden' for its keyword pair so being consistent is nice, and it really is more than just "invisibility" since it's suppressing layout/etc as well; it's not like the content is there but not drawn like "visibility:hidden" does.)

We could then layer a second keyword onto "hidden" to select between the hidden states. Maybe "frozen" for 2 (because it's completely non-interactable), "managed" for 3 (from Fergal's reasoning; it's hidden but I'm managing it, plz send me events and such), and, hm, I can't come up with a better term than "auto" for 4.

So an important question then is, should one of the the hidden states be the default, or should we always require specifying a second keyword? (Similar to the argument of requiring "open"/"closed" to be specified for a shadow root, rather than selecting one by default.)

I could see arguments for 2 being the default (closest to display:none, and conceptually the "simplest" option), or for 4 being the default (the user-friendliest option, which we want it to default to if the author hasn't given a lot of thought to the problem). I think I lean towards 4, which luckily is the one I had trouble with naming in the first place, so we can just drop the second keyword entirely from that state.

So my suggestion ends up being:

```
render-subtree: normal | hidden && [ frozen | managed ]?
```

(Tho wait, don't we have the fifth state of "totally visible, but with the same rendering constraints as when I was hidden", that used to be triggered just by the presence of the attribute? If that's still around, then I suggest "visible" for that state, with no secondary keywords.)

-------

Okay, so "hold-loads" and "hold-upgrades". Whether they're part of render-subtree or a separate property depends on how we see them interacting with changing the values.  Say an author starts an element in "hidden frozen", and wants to hold loads and upgrades; once they're done prepping the subtree and want to switch it to managed/auto, do we expect they'll *still* want to hold loads and upgrades? If the default answer is yes, then these should be controlled by a separate property; if the answer is no or that either is pretty common, we can bundle them in so the author has to re-set them when they change states. Which do y'all expect to be the case?

**Comment by vmpstr** [comment]
I like
```
render-subtree: normal | hidden && [ frozen | managed ]?
```

If there are no objections by `Mon 24 Feb 2020` then I'll change the tokens.

**Comment by fergald** [comment]
Including the state that's mentioned at the end, I think there are 2 states
missing from Tab's list

- Visible and could be auto-hidden by scrolling
- Visible but with the same rendering constraints

**Comment by tabatkins** [comment]
Yeah, the second just means expanding to:

```
render-subtree: normal | visible | hidden && [ frozen | managed ]?
```

What's the first one you mention? Isn't that just `hidden`, but already in view so auto-displayed?

**Comment by fergald** [comment]
It is just `hidden` but currently visible. Is it going to be possible to
know whether something is currently visible or not? Can that be read from
somewhere else?

**Comment by tabatkins** [comment]
When we were doing the attribute, I believe the UA mutated the attribute to make it visible, so you could observe it. That sort of thing is *not* generally done in CSS.

Tho, it might be reasonable to say that this is reflected in the used value of the property? We'd have to define that 'render-subtree' is one of the properties that returns used value from `getComputedStyle()`, then.

Or just define a new API that exposes this info. We already want to do some of these special-purpose "extra layout-ish information that doesn't correspond directly to property values" APIs for Grid and whatnot, so we have some future-precedent to forge here.

**Comment by chrishtr** [comment]
Hmm. We ended up with tokens, but there are only really three rendering states that we've found make any sense: regular rendering, automatic rendering when in viewport, and hidden.

Now that the "UA wants to show your stuff" event is being designed as an independent feature (see #115 ), it makes sense to have a different property that goes with that event. I suggest we discuss the name of this property in #115 .

For this issue, I suggest `subtree-hidden: never | when-offscreen | always`. This reads like plain English that indicates directly what it does. It also avoids using the word `render`, when what's really happening is hiding.

I at first suggested `render-subtree: auto | when-onscreen | never`, but I like this new inversion. Reason being, whether rendering occurs is an implementation detail, and also it allows a future `updateRendering()` method to make sense: it makes sense to render hidden content, but it doesn't make sense to render `render-subtree: never` content.

* `never` is the default, and has no effect
* `when-offscreen` means to have `contain: style layout`, and additionally be `hidden` when the border box of the element does not intersect the visual viewport plus some margin
* `always` means `hidden`, and not visible to UA algorithms unless overridden by some other CSS property designed for that purpose

`hidden` means `contain: style layout size` and subtree is not painted or hit-tested.

In the future, if we want to add `holdloads` or `holdupgrades`, we could make the syntax `subtree-hidden: (never | when-offscreen | always) [holdloads]? [holdupgrades]?`.

**Comment by chrishtr** [comment]
On reflection, maybe the opting into the event has to be done with `subtree-hidden` via another token. Otherwise it seems we end up with scenarios where `subtree-hidden: always` has elements in its subtree that force visibility to UA algorithms, which is bad because it would require style recalc to determine.

New adjusted syntax: `subtree-hidden: (never | when-offscreen | always) [matchable]`

**Comment by vmpstr** [comment]
I like the idea of the property and values being read 'as English' and it makes sense; avoiding the word render also makes sense, since we are speccing this as a strong hint not a requirement that rendering work must be skipped.

However, with the adjusted syntax, we kind of drift away from English readability. For instance,
`subtree-hidden: always hold-upgrades matchable` is not easily understood. In fact, just `subtree-hidden: always matchable` is sounding a bit off to me. Maybe we can iterate to get this to a better state.

I also weakly prefer keeping the property name in the positive (ie "subtree-visibility" or something like that, as opposed "subtree-hidden"). Maybe subtree-state or something like that, although that sounds a bit too general.

----------
I keep going back to Tab's 4 states (which FWIW I agree are the states we care about at this point).

To me they are still the following (and consider reading this with `subtree-visibility` as the property name:
1. `hidden`: "state preserving display: none", no searchability, always hidden
2. `hidden-matchable`: like 1 but fires find-in-page events if matched in the subtree.
3. `managed`: user-agent manages visibility (will be visible on-screen but can be not visible off-screen).
4. `visible`: default

I know managed is a bit contentious in that I mean it's user-agent managed, not developer/script managed. But in lieu of being able to use `auto`, I got nothing better. Have I overlooked something? Do people have objections to this?

When it comes to hold-upgrades / hold-loads, I'm more and more thinking that we will need a separate property for something like that. The reasoning is that `subtree-visilibity` is still a performance hint. It mandates paint and hit-test state. It talks about its interaction with fragment-navigation and find-in-page. It also talks about the notion of being on- or off-screen and how the state changes based on that. It does magical things with containment. It does not (intentionally) talk about performance in a prescriptive way. It only suggests what the user-agent should and shouldn't do. That's also the reason I like the move away from having the word `render` in the property. Going back to hold-upgrades and hold-loads, that seems like a mandatory directive, not a hint. It's script observable and I think if it was just a hint, it would be kind of pointless. So, with that I think we can think about something like `subtree-render-state: auto | hold-upgrades? hold-loads?` (or whatever the formal syntax would be) for the mandated render state. I recommend we consider this in the future, and stop trying to shoehorn the values into this property. Thoughts?

**Comment by vmpstr** [comment]
Going back to `managed` a bit, I would argue that we have a precedent of using `auto` in a non-default way and that's in `overflow`. In fact, if you consider the overflow values we have
`hidden`, `scroll`, `auto`, `visible`.

If `subtree-visibility` becomes `hidden | hidden-matchable | auto | visible` with visible being the default, then it also directly corresponds to `overflow` values <_<

**Comment by vmpstr** [comment]
Also, just to continue redesigning things, because we're on that sailing ship already.

Let's consider again the fact that we're making the matchability signal a separate feature. Here's a wild proposal: why don't we make the matchable a separate property as part of that signal.

`matchable: auto | ignore-visibility`. The latter means that we ignore both `visibility` and `subtree-visibility` when considering whether we can match find-in-page or anchor-link- or scorll-to-text-navigation; if matched, fires an event. (There's a lot of details here, since it's unclear what to do if script reacts to the event or what to do if the script ignores the event, but we need to solve that anyway)

So, `subtree-visibility` becomes `hidden | auto | visible`.

To recap Tab's states:
> 1. normal rendering
> 2. completely hidden; "state-preserving display:none" as Vlad puts it
> 3. hidden but the UA can interact with it and will inform the author when it should be unlocked
> 4. hidden but the UA can interact and auto-unlocks it when necessary

1. `subtree-visibility: visible; matchable: auto`
2. `subtree-visibility: hidden; matchable: auto`
3. `subtree-visibility: hidden; matchable: ignore-visibility`
4. `subtree-visibility: auto; matchable: auto`

For 4, we can make matchable: auto also search auto visibility things...

**Comment by chrishtr** [comment]
I like `subtree-visibility: visible | hidden | auto | hidden-matchable`. In addition to your arguments for why, I also like how it corresponds very closely to the `visibility` CSS property, because its effects regarding hit testing and painting are very similar. The new property just applies to the subtree, not to self.

I don't see a way right now to separate the `matchable` part into its own property, because we'd have to deal with situations like a `matchable: ingnore-visibility` subtree within a `matchable: auto; subtree-visibility: hidden` subtree.

**Comment by vmpstr** [comment]
That's a fair objection to the `matchable` property. I feel like hidden-matchable is a bit of a hack. The good thing is that I think we want to move ahead with `subtree-visibility: visible | hidden | auto` for now. So we can punt on hidden-matchable discussion for a couple of weeks

**Comment by tabatkins** [comment]
I'm generally for Vlad's statement of "prefer positive wording", because double-negatives are harder to parse.

I'm also quite attracted to Vlad's suggestion of matching up with 'overflow' values; the connection does seem pretty intuitive, just about subtree rather than scrollbars.

I'm so-so on 'subtree-visibility' as a name, just because it sounds like a variation of 'visibility' while it actually has a lot of differences. But also: if you ignore the differences and squint a little bit, it does act pretty similar, and uses the same keywords?  So overall 👍 

Also, yeah, separating out the "matchable" and "suppression" qualities into their own properties does sound better to me.

> because we'd have to deal with situations like a matchable: ingnore-visibility subtree within a matchable: auto; subtree-visibility: hidden subtree.

I imagine hidden+auto would be like display:none and suppress the entire subtree, not like 'visibility'.

> So we can punt on hidden-matchable discussion for a couple of weeks

Yup, current syntax design is nicely extensible, let's go for it.

**Comment by fergald** [comment]
I mostly fine with `subtree-visibility: visible | hidden | auto | hidden-matchable` because `hidden` and `hidden-matchable` make a clear distinction between DOM that is hidden because it's not yet "in the document" (we should probably have a good word for this, I don't know what though) and content that is purely being optimized. I'm also not going to argue about syntax of any other version, as long as it expresses this distinction.

What exactly would `subtree-visibility: visible` do? In particular is it the same as `subtree-visibility: auto` when the element is in the viewport? My concern is that if I'm a script trying to do the same thing as `subtree-visibility: auto`, what do I have to do when I make things visible? If it's just flip to `subtree-visibility: visible` that's great. If it's slip to `subtree-visibility: visible` and apply some other styles that seems like a foot-gun, especially if the other styles could evolve moving forward. Put another way, if we don't provide a mode that mean visible just like auto then we will not be able to evolve the meaning of that without breaking script-managed elements.

**Comment by vmpstr** [comment]
`subtree-visibility: visible` is the default state, so for the lack of a better word, it does nothing. Business as usual.

`subtree-visibility: auto` makes the subtree paint in the viewport and not paint outside of the viewport. One important distinction, which is something I think you are talking about here, is that `subtree-visibility: auto` puts in `contain: layout style` always, and `contain: size` if not painted.

We've talked about this on #104 a bit. Basically, I think at this point we don't want `subtree-visibility: visible` to put in 'visible styles' such as containment, mostly because then we need yet another keyword to act as a default. The stance we should take is communication (ie please put in your containment values appropriately if you're planning on managing visibility yourself). This _should_ be sufficient. If there is some concern about it, we also had talk about adding a keyword to `will-change`, something like `will-change: subtree-visibility` which will put in this containment.

**Comment by chrishtr** [comment]
I agree that there is a small risk of problems for developers who toggle between `hidden-matchable` and `visible`, but agree with Vlad that we can address that with `will-change: subtree-visibility` and developer education.

**Comment by tabatkins** [comment]
I'll need to expand on the effects that `will-change` can have (currently it can't apply containment), but that's a very reasonable suggestion for solving the "visible, but with same layout rules" problem.

**Comment by vmpstr** [comment]
This is resolved, AFAIK. I'll open another issue for hidden-matchable or will-change if we need those change or discuss those further.
